common/rate-limiter/src/async_lib.rs

```diff
@@ -0,0 +1,203 @@
+    fn poll_acquire(&mut self, cx: &mut Context<'_>, requested: usize) -> Poll<usize> {
+        loop {
+            // Wait until the delay is finished.
+            if let Some(ref mut delay) = self.delay {
+                ready!(Pin::new(delay).poll(cx));
+                self.delay = None;
+            }
+            // Try to acquire some tokens. If we're rate limited, we have to wait
+            // before trying again.
+            match self.bucket.lock().acquire_tokens(requested) {
+                Ok(allowed) => return Poll::Ready(allowed),
+                Err(wait_time) => {
+                    self.delay = Some(delay_until(tokio::time::Instant::from_std(wait_time)));
+                }
+            }
+        }
+    }
+
+    /// Poll an inner reader or writer, rate limited by the `bucket`.  This will provide an amount
+    /// of bytes allowed ot be read including partial reads.  It will return unused tokens back to
+    /// the bucket on partial reads / writes.
+    pub fn poll_limited<
+        T,
+        Action: FnOnce(Pin<&mut T>, &mut Context<'_>, usize) -> Poll<io::Result<usize>>,
+    >(
+        &mut self,
+        resource: Pin<&mut T>,
+        cx: &mut Context<'_>,
+        requested: usize,
+        poll_resource: Action,
+    ) -> Poll<io::Result<usize>> {
+        let allowed = ready!(self.poll_acquire(cx, requested));
+        let result = poll_resource(resource, cx, allowed);
+
+        // In order to have an accurate throttle rate, we remove tokens, and then add ones we don't use
+        let tokens_to_return = match &result {
+            Poll::Ready(Ok(actual)) => allowed.saturating_sub(*actual),
+            _ => allowed,
+        };
+        self.bucket.lock().return_tokens(tokens_to_return);
+
+        result
+    }
+}
+
+/// A rate limiter for `AsyncRead` or `AsyncWrite` interfaces to rate limit read/write bytes
+///
+/// This will pause and wait to send any future bytes until it's permitted to in the future
+#[pin_project]
+pub struct AsyncRateLimiter<T> {
+    #[pin]
+    inner: T,
+    rate_limiter: PollRateLimiter,
+}
+
+impl<T> AsyncRateLimiter<T> {
+    pub fn new(inner: T, bucket: Option<SharedBucket>) -> Self {
+        Self {
+            inner,
+            rate_limiter: PollRateLimiter::new(bucket),
+        }
+    }
+}
+
+impl<T: AsyncRead> AsyncRead for AsyncRateLimiter<T> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        this.rate_limiter
+            .poll_limited(this.inner, cx, buf.len(), |resource, cx, allowed| {
+                resource.poll_read(cx, &mut buf[..allowed])
+            })
+    }
+}
+
+impl<T: AsyncWrite> AsyncWrite for AsyncRateLimiter<T> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
```

Comment:
why have this as a separate line? same as above

Reply:
I think it was so I didn't type self.project() twice?  Though, it might have been because `self.project()` didn't like being moved twice or something.